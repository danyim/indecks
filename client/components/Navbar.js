import React from 'react';
import { Link, browserHistory } from 'react-router';
import styles from '../styles/components/Navbar';

const Navbar = React.createClass({
  routeParser(path) {
    let addLink, listLink;
    const routeComponents = path.split('/').filter(v => v !== '');

    // console.log('routeComponents', routeComponents);
    if(path === '/') {
      // @/
      addLink = (
        <Link to="/add" title="Add deck"><i className="fa fa-plus-square-o add-icon"></i></Link>
      );
      listLink = (
        //<Link to={`/`} title="View decks"><i className="fa fa-navicon grid-icon"></i></Link>
        <a href="javascript:void(0);" disabled><i className="fa fa-navicon grid-icon"></i></a>
      );
    }
    else if(routeComponents[0] === 'view' && routeComponents.length === 2) {
      // @/view/:deckId
      addLink = (
        <Link to={`/add/${routeComponents[1]}`} title="Add card"><i className="fa fa-plus-square-o add-icon"></i></Link>
      );
      listLink = (
        <Link to={`/`} title="View decks"><i className="fa fa-navicon grid-icon"></i></Link>
      );
    }
    else if(routeComponents[0] === 'add' && routeComponents.length === 2) {
      // @/add/:deckId
      addLink = (
        <Link to={`/add/${routeComponents[1]}`} title="Add card"><i className="fa fa-plus-square-o add-icon"></i></Link>
      );
      listLink = (
        <Link to={`/view/${routeComponents[1]}`} title="View cards"><i className="fa fa-navicon grid-icon"></i></Link>
      );
    }
    else if(routeComponents[0] === 'view' && routeComponents.length === 3) {
      // @/view/:deckId/:cardId
      addLink = (
        <Link to={`/add/${routeComponents[1]}`} title="Add card"><i className="fa fa-plus-square-o add-icon"></i></Link>
      );
      listLink = (
        <Link to={`/view/${routeComponents[1]}`} title="View cards"><i className="fa fa-navicon grid-icon"></i></Link>
      );
    }
    else {
      addLink = (
        <a href="javascript:void(0);" disabled><i className="fa fa-plus-square-o add-icon"></i></a>
      );
      listLink = (
        <a href="javascript:void(0);" disabled><i className="fa fa-navicon grid-icon"></i></a>
      );
    }

    return {
      addLink, listLink
    };
  },
  render() {
    let addLink, listLink;
    const routeParser = this.routeParser;

    browserHistory.listen(function(ev) {
      const { addLink: add, listLink: list } = routeParser(ev.pathname);
      addLink = add;
      listLink = list;
    });

    console.log(styles);
    return (
      <nav className={styles['nav-container']}>
        <div className={styles['nav-outer-container']}>
          <div className={styles['nav-inner-container']}>
            <span className={styles['nav-left']}>
              <div className="rounded-button">
                {addLink}
              </div>
            </span>
            <h1>
              <Link to="/">in<span className={styles['title-bold']}>decks</span></Link>
            </h1>
            <span className={styles['nav-right']}>
              <div className="rounded-button">
                {listLink}
              </div>
            </span>
          </div>
        </div>
      </nav>
    )
  }
});

export default Navbar;
