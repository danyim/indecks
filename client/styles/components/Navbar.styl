.nav-container
  order 1
  height 70px
  background rgba(255,255,255,0.8)
  border-bottom 1px solid lighten(gray, 85%)
  h1
    font 200 2.5rem 'Dosis', sans-serif
    letter-spacing -0.05rem
    // text-shadow 0px 4px 0 rgba(18, 86, 136, 0.11)
  a
    font-weight 300
    color black
    text-decoration none
    &:focus
      outline 0

.nav-outer-container
  flex-shrink 0
  align-items center
.nav-inner-container
  max-width pageWidth
  padding 20px 20px
  width 100%
  justify-content space-between
  flex-direction row
  h1
    flex-shrink 0
.nav-left, .nav-right
  font-size 1.5rem
.title-bold
  font-weight 400
