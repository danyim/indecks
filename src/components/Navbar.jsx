import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import ModalTypes from './ModalTypes'
import styles from '../styles/components/Navbar.styl'

class Navbar extends React.Component {
  static propTypes = {
    changeActiveModal: PropTypes.func.isRequired,
    location: PropTypes.shape({
      pathname: PropTypes.string.isRequired,
    }).isRequired,
  }

  static defaultProps = {}

  static renderLink(url, title, faClassName) {
    return (
      <Link to={url} title={title}>
        <i className={`fa ${faClassName}`} />
      </Link>
    )
  }

  static renderModalLink(openModalFn, faClassName) {
    return (
      <a onClick={openModalFn} className="pointer" role="presentation">
        <i className={`fa ${faClassName}`} />
      </a>
    )
  }

  static renderNoAction(faClassName = 'fa-plus-square-o') {
    return (
      <a disabled>
        <i className={`fa ${faClassName}`} />
      </a>
    )
  }

  constructor(props) {
    super(props)
    this.routeParser = this.routeParser.bind(this)
  }

  routeParser(path) {
    let addLink
    let listLink
    const routeComponents = path.split('/').filter(v => v !== '')

    if (path === '/') {
      // @/
      // addLink = Navbar.renderLink('/add', 'Add deck', 'fa-plus-square-o');
      addLink = Navbar.renderModalLink(
        () => this.props.changeActiveModal(ModalTypes.Import),
        'fa-plus-square-o'
      )
      listLink = Navbar.renderModalLink(
        () => this.props.changeActiveModal(ModalTypes.Settings),
        'fa-cog'
      )
      // listLink = Navbar.renderLink('/settings', ModalTypes.Settings, 'fa-cog');
      // <a href="javascript:void(0);" disabled><i className="fa fa-navicon"></i></a>
    } else if (routeComponents[0] === 'view' && routeComponents.length === 2) {
      // @/view/:deckId
      addLink = Navbar.renderLink(
        `/add/${routeComponents[1]}`,
        'Add card',
        'fa-plus-square-o'
      )
      listLink = Navbar.renderLink('/', 'View decks', 'fa-clone')
    } else if (routeComponents[0] === 'add' && routeComponents.length === 2) {
      // @/add/:deckId
      addLink = Navbar.renderLink(
        `/add/${routeComponents[1]}`,
        'Add card',
        'fa-plus-square-o'
      )
      listLink = Navbar.renderLink(
        `/view/${routeComponents[1]}`,
        'View cards',
        'fa-square-o'
      )
    } else if (routeComponents[0] === 'view' && routeComponents.length === 3) {
      // @/view/:deckId/:cardId
      addLink = Navbar.renderLink(
        `/add/${routeComponents[1]}`,
        'Add card',
        'fa-plus-square-o'
      )
      listLink = Navbar.renderLink(
        `/view/${routeComponents[1]}`,
        'View cards',
        'fa-square-o'
      )
    } else if (routeComponents[0] === 'add' && routeComponents.length === 1) {
      // @/add
      addLink = Navbar.renderNoAction()
      listLink = Navbar.renderLink('/', 'View decks', 'fa-th')
    } else if (routeComponents[0] === 'edit' && routeComponents.length === 3) {
      // @/edit/:deckId/:cardIndex
      addLink = Navbar.renderNoAction()
      listLink = Navbar.renderLink(
        `/view/${routeComponents[1]}`,
        'View cards',
        'fa-square-o'
      )
    } else {
      addLink = Navbar.renderModalLink(
        () => this.props.changeActiveModal(ModalTypes.Import),
        'fa-plus-square-o'
      )
      listLink = null
      // listLink = (
      //   <a href="javascript:void(0);" disabled><i className="fa fa-navicon"></i></a>
      // );
    }

    return {
      addLink,
      listLink,
    }
  }

  render() {
    const { addLink, listLink } = this.routeParser(this.props.location.pathname)

    return (
      <nav className={`${styles['nav-container']}`}>
        <div className={`${styles['nav-outer-container']}`}>
          <div className={`${styles['nav-inner-container']}`}>
            <span className={`${styles['nav-left']}`}>
              <div className={`${styles['rounded-button']}`}>{addLink}</div>
            </span>
            <h1>
              <Link to="/">
                in<span className={`${styles['title-bold']}`}>decks</span>
              </Link>
            </h1>
            <span className={`${styles['nav-right']}`}>
              <div className={`${styles['rounded-button']}`}>{listLink}</div>
            </span>
          </div>
        </div>
      </nav>
    )
  }
}

export default Navbar
