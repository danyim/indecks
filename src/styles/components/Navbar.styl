@import '../_variables';

.nav-container
  order: 1
  height: 70px
  background: rgba(255, 255, 255, .8)
  border-bottom: 1px solid lighten(gray, 85%)
  h1
    font: 200 2.5rem 'Dosis', sans-serif
    letter-spacing: -.05rem
    // text-shadow 0px 4px 0 rgba(18, 86, 136, 0.11)
  a
    font-weight: 300
    color: black
    text-decoration: none
    &:focus
      outline: 0

.nav-outer-container
  flex-shrink: 0
  align-items: center
.nav-inner-container
  max-width: pageWidth
  padding: 2.0rem 1.0rem
  width: 100%
  justify-content: space-between
  flex-direction: row
  h1
    flex-shrink: 0
.title-bold
  font-weight: 400

.nav-left
  font-size: 2.3rem
.nav-right
  font-size: 2.2rem

.rounded-button
  border: 1px solid lightgrey
  border-radius: .3rem
  padding: .3rem .7rem
  a
    color: black
  a[disabled]
    color: gray
    cursor: not-allowed
  a.hidden
    visibility: hidden

.modal-overlay
  background-color: rgba(0, 0, 0, .15)
  position: fixed
  top: 0
  left: 0
  right: 0
  bottom: 0

.modal-content
  width: 75%
  max-width: 700px
  background-color: white
  position: absolute
  top: 50%
  left: 50%
  right: auto
  bottom: auto
  margin-right: -50%
  transform: translate(-50%, -50%) !important
  border: 1px solid black
  padding: 4rem
  transition: all .2s

@media (max-width: cardWidth)
  .modal-content
    width: 95%
    position: inherit
    top: 2%
    left: 50%
    transform: translate(-50%, 1%) !important
    padding: 2rem
    overflow: scroll
    height: 95%

@media (max-width: cardWidth)
  .nav-container, .nav-outer-container
    height: 45px
  .nav-inner-container
    padding: .5rem 1.0rem
  .rounded-button
    border: 0
    margin-top: .3rem
