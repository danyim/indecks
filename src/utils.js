export const createReducer = (initialState, handlers) => {
  return function reducer(state = initialState, action = {}) {
    if (handlers.hasOwnProperty(action.type)) {
      return handlers[action.type](state, action);
    }
    return state;
  };
};

/**
 * Given an array of consecutive numbers, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a Array of numbers
 * @return {Number}   Missing value
 */
export const findMissing = (a) => {
  let expectedNextValue;
  for(let k of a)
  {
    expectedNextValue = a[k] + 1;
    if(k <= a.length - 1 && expectedNextValue !== a[k + 1]) {
      return expectedNextValue;
    }
  }
  return null;
};

/**
 * Given an array with potentially missing values at parameter key, find the
 * next best 0-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) => {
  return findMissing(a.map(x => x[key])) || Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1);
};

