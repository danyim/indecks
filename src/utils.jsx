import React from 'react'
import { Route } from 'react-router-dom'
import _ from 'lodash'

export const createReducer = (initialState, handlers) =>
  function reducer(state = initialState, action = {}) {
    if (Object.prototype.hasOwnProperty.call(handlers, action.type)) {
      return handlers[action.type](state, action)
    }
    return state
  }

/**
 * Given an array of consecutive numbers from 1 to n, find the missing value.
 * Returns null if no missing value found
 * @param  {Array}  a   Array of numbers
 * @param  {Number} min Minimum value (0 or 1)
 * @return {Number}     Missing value
 */
export const findMissing = (a, min = 1) => {
  let expected = min
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== expected) {
      return expected
    }
    expected++
  }
  return null
}

/**
 * Given an array of objects, check the value of the key and return the
 * best 1-based index to insert the next value
 * @param  {Array}  a   An array of objects
 * @param  {String} key Key to look for
 * @return {Number}     Index of the best next value
 */
export const findBestNextIndex = (a, key) =>
  findMissing(a.map(x => x[key])) ||
  Math.max(0, Math.max.apply(null, a.map(x => x[key])) + 1)

/**
 * Helper function to generate routes
 * @param  {Object} route) Route object that takes the form
 *                         {
 *                           exact: boolean,
 *                           path: string,
 *                           component: Component,
 *                           routes: Routes
 *                         }
 * @return {Component}        Returns a <Route> component to plug into a <Switch>
 */
export const routeWithSubRoutes = route =>
  <Route
    key={_.uniqueId()}
    exact={route.exact || false}
    path={route.path}
    render={props =>
      // Pass the sub-routes down to keep nesting
      // <route.component {...props}>
      //   { route.children ? route.children.map(child => routeWithSubRoutes(child)) : null }
      // </route.component>
      <route.component {...props} routes={route.routes} />}
  />
