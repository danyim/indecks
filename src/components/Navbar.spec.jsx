import toJson from 'enzyme-to-json'
import { shallow, mount, setup, setupFull } from '../test/utils' // eslint-disable-line no-unused-vars
import Navbar from './Navbar'

describe('Navbar', () => {
  let props

  beforeEach(() => {
    props = {
      changeActiveModal: () => {},
      location: {
        pathname: '/'
      }
    }
  })

  it('should render self and subcomponents', () => {
    const wrapper = shallow(Navbar, props)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
