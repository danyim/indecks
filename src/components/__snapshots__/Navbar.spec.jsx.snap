// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render self and subcomponents 1`] = `
<nav
  className="nav-container"
>
  <div
    className="nav-outer-container"
  >
    <div
      className="nav-inner-container"
    >
      <span
        className="nav-left"
      >
        <div
          className="rounded-button"
        >
          <a
            className="pointer"
            onClick={[Function]}
            role="presentation"
          >
            <i
              className="fa fa-plus-square-o"
            />
          </a>
        </div>
      </span>
      <h1>
        <Link
          replace={false}
          to="/"
        >
          in
          <span
            className="title-bold"
          >
            decks
          </span>
        </Link>
      </h1>
      <span
        className="nav-right"
      >
        <div
          className="rounded-button"
        >
          <a
            className="pointer"
            onClick={[Function]}
            role="presentation"
          >
            <i
              className="fa fa-cog"
            />
          </a>
        </div>
      </span>
    </div>
  </div>
</nav>
`
